code-monkey/src/agents/agent.py

```diff
@@ -1,78 +1,148 @@
     def initialize(self):
         pass
 
-    def run_prompt(self, prompt: str):
+    @instrument("Agent.run_prompt", ["prompt"])
+    async def run_prompt(self, prompt: str):
         import logging
         logger = logging.getLogger(__name__)
         logger.debug(f"Running prompt: {prompt}")
-        result = self.model.run_prompt(prompt)
-        logger.debug(f"Prompt result: {result}")
-        return result
+
+        system = SystemMessage(content=self.get_system_prompt())
+        msg = HumanMessage(content=self.prepare_prompt(prompt))
+
+        modified_files: Set[str] = set()
+
+        async for event in self.model.astream_events({
+            "messages": [system, msg],
+        }, self.config, version="v2"):
+            kind = event["event"]
+            if kind == "on_chat_model_start":
+                # for msg in event["data"]["input"]["messages"][0]:
+                #     print("----")
+                #     print(msg.content)
+                continue
+
+            if kind == "on_chat_model_stream":
+                # NB(toshok) we can stream text as it comes in by doing it here,
+                # but I'm doing the entire thing in one go in the
+                # on_chat_model_end event just below.
+                continue
+
+            if kind == "on_tool_start":
+                logger.debug("----")
+                tool_name = event["name"]
+                tool_input = event["data"]["input"]
+                logger.debug("tool start:")
+                logger.debug(f"      name: {tool_name}")
+                logger.debug(f"     input: {repr(tool_input)}")
+                continue
+
+            if kind == "on_tool_end":
+                logger.debug("----")
+                tool_name = event["name"]
+                tool_output = event["data"]["output"].content[:20]
+                logger.debug("tool end:")
+                logger.debug(f"    name: {tool_name}")
+                logger.debug(f"  output: {repr(tool_output)}")
+                continue
+
+            if kind == "on_chat_model_end":
+                logger.debug("----")
+                # anthropic seems give us content as a list?
+                if isinstance(event["data"]["output"].content, list):
+                    print(event["data"]["output"].content[0]["text"])
+                else:
+                    print(event["data"]["output"].content)
+                # TODO(toshok) still need to accumulate tokens
+                continue
+
+            if kind == "on_custom_event":
+                # TODO(toshok) a better dispatch mechanism would be nice, but
+                # there's only one event type currently
+                if event["name"] == "file_modified":
+                    modified_files.add(event["data"])
+                continue
+
+        self.handle_completion(
+            modified_files,
+        )
+
+        # TODO(toshok) we aren't returning a result here, and likely should...
```

Comment:
The agent should generally return a final report, maybe the last `text` that we received. We can fix that later, once we get into multi-agent stuff.

Reply:
right.  also unclear how we want to deal with responses from agents - they don't have to be the result of a prompt - they can instead be events the tool emits.